src/ReturnTypes/EnumerableGroupByReturnTypeExtension.php

```diff
@@ -0,0 +1,148 @@
+        $objectClassReflections = $calledOnType->getObjectClassReflections();
+
+        if (!isset($objectClassReflections[0])) {
+            return null;
+        }
+
+        $groupByType = $scope->getType($methodCall->getArgs()[0]->value);
+        $propertyTypes = [];
+
+        $collectionName = $objectClassReflections[0]->getName();
+
+        $valueType = new MixedType();
+
+        if ($objectClassReflections[0]->isGeneric()) {
+            $tValueType = $methodReflection->getDeclaringClass()->getActiveTemplateTypeMap()->getType('TValue');
+
+            if ($tValueType !== null) {
+                $valueType = $tValueType;
+            }
+        }
+
+        if ($groupByType->isString()->yes()) {
+            $propertyTypes[] = $groupByType;
+        } elseif ($groupByType->isConstantArray()->yes()) {
+            if ($groupByType->isIterableAtLeastOnce()->no()) {
+                return $this->unknownKeysType($collectionName, $valueType);
+            }
+
+            $valuesArray = $groupByType->getConstantArrays()[0]->getValueTypes();
+
+            foreach ($valuesArray as $valuesType) {
+                $propertyTypes[] = $valuesType;
+            }
+        } else {
+            return $this->unknownKeysType($collectionName, new MixedType());
+        }
+
+
+        $innerKeyType = new IntegerType();
+
+        if (count($methodCall->getArgs()) >= 2) {
+            $preserveType = $scope->getType($methodCall->getArgs()[1]->value);
+            $tKeyType = $methodReflection->getDeclaringClass()->getActiveTemplateTypeMap()->getType('TKey');
+
+            if ($tKeyType === null) {
+                $innerKeyType = new BenevolentUnionType([new IntegerType(), new StringType()]);
+            } elseif ($preserveType->isTrue()->yes()) {
+                $innerKeyType = $tKeyType;
+            } elseif ($preserveType->isTrue()->maybe()) {
+                $innerKeyType = TypeCombinator::union($tKeyType, new IntegerType());
+            }
+        }
+
+
+        $inner = new GenericObjectType($collectionName, [$innerKeyType, $valueType]);
+
+        foreach (array_reverse($propertyTypes) as $propertyType) {
+            if (count($propertyType->getConstantStrings()) > 0) {
+                $types = [];
+                foreach ($propertyType->getConstantStrings() as $constantString) {
+                    if ($valueType->hasProperty($constantString->getValue())->yes()) {
+                        $types[] = $valueType->getProperty($constantString->getValue(), new OutOfClassScope())->getReadableType();
+                    } elseif ($valueType->hasOffsetValueType($constantString)->yes()) {
+                        $types[] = $valueType->getOffsetValueType($constantString);
+                    }
+                }
+
+                if (count($types) === 0) {
+                    $keyType = new ConstantStringType('');
+                } else {
+                    $keyType = TypeCombinator::union(...$types);
```

Comment:
Shouldn't we check if the types are either `int` or `string` ? What I mean is that for example objects can't be array keys. Or if any of the types were nullable we would have for example `int|null` as collection key. Which does not make sense.

Maybe we can call `toArrayKey` on this type?